### ISSUE TITLE
How to use fairness metrics for regression problem

### ISSUE THREAD

**Issue by pradeepdev-1995** [fairlearn/fairlearn#1217]
I could observe that fairness metrics such as ```equalized_odds_ratio, equalized_odds_difference,  demographic_parity_ratio, demographic_parity_difference```, are expecting **y_true** and **y_pred** as arguments .  In the source code, it shows that **y_pred** should be the output of a **classifier**. 
```
y_pred : array-like
        Predicted labels :math:`h(X)` returned by the classifier.
```
So these metrics are only for evaluating classification problems? If so which evaluation metrics I should use for my regression use case? please share the example

**Comment by romanlutz** [comment]
Yes they're for classification only.

You can create a metric that makes sense for your application context and plug it into make_derived_metric like this: https://fairlearn.org/main/user_guide/assessment/custom_fairness_metrics.html

MAE or MSE are examples there, but you should consider your application and decide if they are useful there. Just because DP exists for binar classification doesn't make it useful in every case. It depends on the harms your application might cause. For more please refer to https://fairlearn.org/main/user_guide/assessment/perform_fairness_assessment.html

**Comment by pradeepdev-1995** [comment]
@romanlutz Please provide any existing example code or link for derived metric for regression.

The link you provided is derived metric for classification.

By reading this I am not getting clear picture to create a derived fairness metric for regression.

**Comment by pradeepdev-1995** [comment]
@romanlutz I tried in the following way and got two values
```
from fairlearn.metrics import make_derived_metric, MetricFrame
from sklearn.metrics import mean_squared_error
y_true = [10.1, 11.2, 12.3, 13.4, 14.5, 15.6]
y_pred = [10.2, 12.4, 12.3, 12.4, 17.9, 16.3]
sf_data = ['b', 'b', 'a', 'b', 'b', 'c']
recall_difference = make_derived_metric(metric=mean_squared_error,transform='difference')
print(recall_difference(y_true, y_pred,sensitive_features=sf_data))
//3.502499999999998
print(MetricFrame(metrics=mean_squared_error,y_true=y_true,y_pred=y_pred,sensitive_features=sf_data).difference())
//3.502499999999998
```
Is this the correct way?

**Comment by pradeepdev-1995** [comment]
@romanlutz if I go with **MSE difference** as a potential fairness metric then what will be the optimal value of a regression model to be fair? hope it is **0**.
similarly, if I go with the **MSE ratio** as a potential fairness metric then I think **1**  will be the optimal value of a regression model to be fair.

**Comment by romanlutz** [comment]
Theoretically, yes, the optimal value for MSE difference is 0 and for the ratio it's 1. I would very much like to point out that if you're purely going for optimization of metrics then you're very very likely doing something wrong. While fairness metrics can provide valuable insights, it's essential to understand that fairness is a sociotechnical concept that may not be fully captured by metrics alone. In other words, it's crucial to have a deep understanding of the application context and how the model is used in practice to determine whether it is fair. For example, a model may meet certain fairness metrics but still have unintended consequences that only become apparent in the context of real-world use. It's also worth noting that different stakeholders may have different definitions of fairness depending on their values and priorities. Therefore, a collaborative effort is necessary to define what fairness means in a specific application context and ensure that the model meets those criteria. While fairness metrics are a valuable tool, they should not be the only factor in determining whether a machine learning model is fair.

I hope this response provides helpful insights. Please let me know if you have any further questions or concerns.

**Comment by pradeepdev-1995** [comment]
@romanlutz I am following in this way
step:1 
calculate fairness in the original dataset using **MSE ratio**
```mse_ratio = make_derived_metric(metric=mean_absolute_error,transform='ratio')
mse_ratio(y_true, y_pred,sensitive_features=sf_data)
```
step:2
Apply mitigation over the regression model using **GridSearch** or **ExponentiatedGradient** with **BoundedGroupLoss**
```
constraint = BoundedGroupLoss(SquareLoss(), upper_bound=0.1)
mitigator = ExponentiatedGradient(regr_model, constraint)
mitigator.fit(X, y_true, sensitive_features=sf_data)
y_pred_mitigated = mitigator.predict(X)
```
Step-3
calculate fairness in the y_pred_mitigated target variables using **MSE ratio** again like in step one. And check whether MSE ratio moved to 1 or not

Hope this method will work. Let me know if my method is wrong
